# Домашнее задание 1

In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [2]:
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [3]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1,  1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])
X.shape

(2, 10)

In [4]:
def calc_mse(y_pred, y):
    return np.mean((y - y_pred) ** 2)

1. Подберите скорость обучения (alpha) и количество итераций:

In [5]:
n = X.shape[1]
alpha = 1e-2

Для того, чтобы найти лучшую скорость обучения и количество итераций, создадим цикл, в котором и будем определять необходимое количество итераций для данной скорости обучения (_alpha_). В качестве критерия остановки будем использовать целевую разницу между предыдущим и текущим значением `mse`.

In [6]:
W = np.array([1, 0.5])
target_diff = 1e-7
diff = np.inf
i = 0
while target_diff < diff:
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    new_err = calc_mse(y, np.dot(W, X))
    diff = np.abs(err - new_err)
    i += 1
    if i % 100 == 0:
        print(i, W, err, diff)
print(i, W, err, diff)

100 [31.72648327  6.77395294] 177.92494760058077 2.7304890205622314
200 [41.78014899  4.91746666] 62.255347042578386 0.337575299764886
300 [45.3151501   4.26470166] 47.95490496498401 0.04173504531722472
400 [46.55810299  4.0351809 ] 46.18691542730009 0.005159779192489111
500 [46.99514157  3.95447839] 45.968335680716045 0.0006379128406663881
600 [47.14881009  3.92610232] 45.94131227101915 7.886631911446784e-05
700 [47.20284196  3.91612492] 45.93797131796627 9.750385764561997e-06
800 [47.22184028  3.91261674] 45.93755826989325 1.2054578704123742e-06
900 [47.22852033  3.91138321] 45.93750720401236 1.4903292822054937e-07
920 [47.22920369  3.91125703] 45.93750474242528 9.810889878281159e-08


Таким образом подобраны оптимальные значения c заданной точностью `mse`:
* скорость обучения `alpha` = 1e-2;
* количество итераций - 920. Округлим в большую сторону - 1000.

In [7]:
# подставим найденные значения в заданный цикл
W = np.array([1, 0.5])
for i in range(1000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    if i % 100 == 0:
        print(i, W, err)
print(i, W, err)

0 [2.102 3.9  ] 3173.15
100 [31.88770806  6.74418155] 175.19445858001853
200 [41.83683774  4.90699865] 61.9177717428135
300 [45.33508261  4.26102097] 47.913169919666785
400 [46.56511152  4.03388672] 46.181755648107604
500 [46.99760587  3.95402334] 45.96769776787538
600 [47.14967657  3.92594232] 45.941233404700036
700 [47.20314662  3.91606866] 45.93796156758051
800 [47.2219474   3.91259695] 45.93755706443538
900 [47.228558    3.91137626] 45.937507054979434
999 [47.23086913  3.91094949] 45.93750089064509


*2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [8]:
for i in range(1000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    '''for ii in range(W.shape[0]):
    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))'''
    # W -= (alpha * (1/n * 2 * np.sum(X * (y_pred - y))))
    W -= alpha * (1/n * 2 * X.dot(y_pred - y))
    if i % 100 == 0:
        print(i, W, err)

0 [47.23088237  3.91094704] 45.937500872219864
100 [47.23169965  3.91079613] 45.93750010783411
200 [47.23198702  3.91074306] 45.93750001333172
300 [47.23208806  3.9107244 ] 45.93750000164824
400 [47.23212359  3.91071784] 45.937500000203784
500 [47.23213608  3.91071554] 45.937500000025196
600 [47.23214048  3.91071473] 45.93750000000313
700 [47.23214202  3.91071444] 45.93750000000038
800 [47.23214256  3.91071434] 45.93750000000003
900 [47.23214275  3.9107143 ] 45.937500000000014
